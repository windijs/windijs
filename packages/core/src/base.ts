import { resetMeta, SymbolProxy } from "@windijs/helpers";

import { guard } from "./api";

import type { Handler } from "@windijs/helpers";

export class Utility<T extends object = {}> implements ProxyHandler<T> {
  uid: string;
  plugins: ((p: string) => any)[];

  constructor(uid: string) {
    this.uid = uid;
    this.plugins = [];
  }

  get(target: T, prop: string | symbol) {
    if (Reflect.has(target, prop)) return Reflect.get(target, prop);
    resetMeta(this.uid);
    let result;
    for (const plugin of this.plugins) {
      result = plugin(prop as string);
      if (result) return result;
    }
  }

  set(target: T, prop: string | symbol, value: any) {
    return Reflect.defineProperty(target, prop, { value, writable: true });
  }

  public case<K extends string, U>(trigger: K, plugin: Handler<U>): Utility<T & { [P in K]: U }> {
    this.plugins.push(guard(trigger, plugin).get);
    return this as unknown as Utility<T & { [P in K]: U }>;
  }

  public use<U>(plugin: Handler<U>): Utility<T & U> {
    this.plugins.push(plugin.get);
    return this as unknown as Utility<T & U>;
  }

  public init(): T;
  public init<F extends Function | object>(target: F): F & T;
  public init<F extends Function | object>(target: F, handler: ProxyHandler<F>): F & T;
  public init<F extends Function | object>(target?: F | T, handler?: ProxyHandler<F>) {
    if (!target) target = function () {} as T;
    Object.defineProperty(target, SymbolProxy, { value: true });
    return new Proxy(target, handler || this);
  }
}

/**
 * Create a new utility.
 * @param uid Utility ID, usually it should be consistent with the variiable name you declared. Such as, `const bg = createUtility("bg")`
 * @returns Utility
 */
export function createUtility(uid: string) {
  return new Utility(uid);
}
