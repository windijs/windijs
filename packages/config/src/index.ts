export * from "./bg";
export * from "./svg";
export * from "./flex";
export * from "./grid";
export * from "./border";
export * from "./effects";
export * from "./behaviors";
export * from "./filter";
export * from "./opacity";
export * from "./sizing";
export * from "./spacing";
export * from "./animation";
export * from "./transform";
export * from "./positioning";
export * from "./interactivity";
export * from "./transitions";
export * from "./typography";
export * from "./keyframes";
export * from "./interface";
