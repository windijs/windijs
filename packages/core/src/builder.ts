import { css, getFirstVar, hexToRGB, prop, sliceColor, SymbolCSS } from "@windijs/helpers";
import { dashToCamel, parenWrap } from "@windijs/shared";

import { prefixImageRendering, prefixKeyframes, prefixNotHidden, prefixPlaceholder, prefixWritingMode } from "./prefixer";

import type {
  ColorStyleObject,
  CSSAngle,
  CSSLinearColorStopOrHint,
  CSSMap,
  CSSObject,
  CSSPrefixer,
  CSSSideOrCorner,
  StyleObject,
  StyleProperties,
} from "@windijs/helpers";

export const buildProperty = (property: StyleProperties | StyleProperties[], value: string) =>
  (Array.isArray(property) ? Object.fromEntries(property.map(i => [i, value as string])) : { [property]: value as string }) as CSSObject;

export function buildStatic(property: StyleProperties | StyleProperties[], value: unknown): StyleObject | undefined {
  if (typeof value === "string") return css(buildProperty(property, value));
}

// TODO: consider put this into api, allow user to define their methods
export function buildColor(
  colorProperty: StyleProperties | StyleProperties[],
  colorOpacityProperty: undefined,
  value: unknown,
  prefixer?: (css: CSSObject) => CSSObject
): StyleObject;
export function buildColor(
  colorProperty: StyleProperties | StyleProperties[],
  colorOpacityProperty: string,
  value: unknown,
  prefixer?: (css: CSSObject) => CSSObject
): ColorStyleObject;
export function buildColor(
  colorProperty: StyleProperties | StyleProperties[],
  colorOpacityProperty: string | undefined,
  value: unknown,
  prefixer: CSSPrefixer = i => i
): StyleObject | ColorStyleObject | undefined {
  if (typeof value !== "string") return undefined;

  let decl = buildProperty(colorProperty, value);

  if (colorOpacityProperty != null) {
    if (value.startsWith("#")) {
      const [r, g, b, a] = hexToRGB(value);
      decl = {
        ...buildProperty(colorProperty, parenWrap("rgba", [r, g, b, parenWrap("var", colorOpacityProperty)].join(", "))),
        [colorOpacityProperty]: a.toString(),
      };
    } else if (/^(rgb|hwb|hsl)/.test(value)) {
      const values = sliceColor(value);
      decl = {
        ...buildProperty(
          colorProperty,
          value.startsWith("hwb")
            ? parenWrap("hwb", values.slice(0, 3).join(" ") + " / " + parenWrap("var", colorOpacityProperty))
            : parenWrap(value.startsWith("hsl") ? "hsla" : "rgba", [...values.slice(0, 3), parenWrap("var", colorOpacityProperty)].join(", "))
        ),
        [colorOpacityProperty]: values[3] ?? "1",
      };
    }

    const gradient = (obj: CSSObject) => ({
      get gradient() {
        return css(
          prefixer({
            ...obj,
            backgroundImage: "linear-gradient(180deg, rgba(255, 255, 255, 0.15), rgba(255, 255, 255, 0))",
          })
        );
      },
    });

    return css(prefixer(decl), {
      opacity(op: number) {
        const obj: CSSObject = { ...decl };
        obj[colorOpacityProperty] = (op / 100).toString();
        return css(prefixer(obj), gradient(obj));
      },
      ...gradient(decl),
    }) as ColorStyleObject;
  }
  return css(prefixer(decl));
}

export function buildContainer<T extends Record<string, string | [string, CSSObject]>>(screens: T, center = false) {
  const m = new Map() as CSSMap;
  m.set("width", "100%");
  if (center) m.set("marginLeft", "auto") && m.set("marginRight", "auto");

  for (const [k, v] of Object.entries(screens))
    if (k === "DEFAULT") {
      if (Array.isArray(v)) Object.entries(v[1]).forEach(([k, v]) => m.set(k, v));
    } else if (typeof v === "string") m.set(`@media (min-width: ${v})`, { maxWidth: v });
    else m.set(`@media (min-width: ${v[0]})`, { maxWidth: v[0], ...v[1] });

  return css(m);
}

export function buildNotHidden(property: StyleProperties | StyleProperties[]) {
  return (v: unknown) => css(prefixNotHidden(buildProperty(property, v as string)));
}

export function buildFontSize(fontSize: string, lineHeight?: string, others?: { [key in StyleProperties]: string }): StyleObject {
  let decl: CSSObject = { fontSize };
  if (lineHeight != null) decl.lineHeight = lineHeight;
  if (others != null) decl = { ...decl, ...others } as CSSObject;

  return css(decl);
}

export function buildFlexDirection(v: unknown) {
  if (typeof v !== "string") return undefined;
  return css({
    "-webkit-box-orient": v.startsWith("row") ? "horizontal" : "vertical",
    "-webkit-box-direction": v.includes("reverse") ? "reverse" : "normal",
    "-ms-flex-direction": v,
    "-webkit-flex-direction": v,
    flexDirection: v,
  });
}

export function buildFlexStretch(v: unknown) {
  if (typeof v !== "string") return undefined;
  return css({
    "-webkit-box-flex": v === "none" || v.startsWith("0") ? "0" : "1",
    "-ms-flex": v,
    "-webkit-flex": v,
    flex: v,
  });
}

export function buildReverse(v: unknown, first: StyleProperties, second: StyleProperties, rev: string): CSSObject {
  return {
    [rev]: "0",
    [first]: `calc(${v} * var(${rev}))`,
    [second]: `calc(${v} * calc(1 - var(${rev})))`,
  };
}

export function buildDivideY(v: unknown) {
  return css(prefixNotHidden(buildReverse(v, "borderBottomWidth", "borderTopWidth", "--w-divide-y-reverse")));
}

export function buildDivideX(v: unknown) {
  return css(prefixNotHidden(buildReverse(v, "borderRightWidth", "borderLeftWidth", "--w-divide-x-reverse")));
}

export function buildSpaceBetweenY(v: unknown) {
  return css(prefixNotHidden(buildReverse(v, "marginBottom", "marginTop", "--w-space-y-reverse")));
}

export function buildSpaceBetweenX(v: unknown) {
  return css(prefixNotHidden(buildReverse(v, "marginRight", "marginLeft", "--w-space-x-reverse")));
}

export const joinFilters = (filters: (string | StyleObject)[]) => filters.map(i => (typeof i === "string" ? i : getFirstVar(i)?.[1])).join(" ");

export const buildFilter = (...filters: (string | StyleObject)[]) =>
  css(Object.fromEntries(["-webkit-filter", "filter"].map(i => [i, joinFilters(filters)])));

export const buildBackdropFilter = (...filters: (string | StyleObject)[]) =>
  css(Object.fromEntries(["-webkit-backdrop-filter", "backdropFilter"].map(i => [i, joinFilters(filters)])));

export const joinTransforms = (transformations: (string | StyleObject)[]) =>
  transformations
    .map(t => {
      if (typeof t === "string") return t;
      const [k, v] = getFirstVar(t) ?? [];
      return k && v ? parenWrap(dashToCamel(k.slice(4)), v) : "";
    })
    .join(" ");

export const buildTransform = (...transformations: (string | StyleObject)[]) =>
  css(Object.fromEntries(["-webkit-transform", "-ms-transform", "transform"].map(i => [i, joinTransforms(transformations)])));

export function buildTransition(property: string, ...styles: StyleObject[]): StyleObject {
  return css(Object.assign({ transitionProperty: property }, ...styles.map(i => i[SymbolCSS])));
}

export function buildKeyframes(name: string, keyframes: Record<string, CSSObject>): CSSObject {
  keyframes = Object.fromEntries(Object.entries(keyframes).map(([k, v]) => [k, prefixKeyframes(v)]));
  return {
    [`@-webkit-keyframes ${name}`]: keyframes,
    [`@keyframes ${name}`]: keyframes,
  };
}

function buildODir(v: string): string {
  return [
    v.includes("top") ? "bottom" : v.includes("bottom") ? "top" : null,
    v.includes("left") ? "right" : v.includes("right") ? "left" : null,
    v.endsWith("deg") ? (Math.abs(450 - +v.slice(0, -3)) % 360) + "deg" : null,
  ]
    .filter(i => i != null)
    .join(" ");
}

function buildWebkitGradient(v: string, colorStops: CSSLinearColorStopOrHint[]): string {
  const len = colorStops.length;
  const dirs = v.split(" ");
  const stops = colorStops.map((stop, i) =>
    i === 0 && typeof stop === "string"
      ? parenWrap("from", stop)
      : i === len - 1 && len > 1 && typeof stop === "string"
      ? parenWrap("to", stop)
      : parenWrap("color-stop", Array.isArray(stop) ? stop.reverse().join(", ") : stop.toString())
  );
  return `-webkit-gradient(linear, ${dirs.find(v => v === "left") ? "right" : "left"} ${dirs.find(v => v === "top") ? "bottom" : "top"}, ${
    dirs.find(v => v === "left" || v === "right") ?? "left"
  } ${dirs.find(v => v === "top" || v === "bottom") ?? "top"}, ${stops.join(", ")})`;
}

export function buildGradientDirection(v: unknown, colorStops: CSSLinearColorStopOrHint[] = ["var(--w-gradient-stops)"]) {
  if (typeof v !== "string") return;
  let d;
  const isTo = () => /top|left|bottom|right/.test(v as string);
  if (isTo()) d = "to " + v;
  else {
    d = v;
    switch (v) {
      case "0deg":
        v = "top";
        break;
      case "90deg":
        v = "right";
        break;
      case "180deg":
        v = "bottom";
        break;
      case "270deg":
        v = "left";
        break;
    }
  }
  const c = colorStops.map(i => (Array.isArray(i) ? i.join(" ") : i)).join(", ");
  const g = `linear-gradient(${d}, ${c})`;
  const o = `-o-linear-gradient(${buildODir(v as string)}, ${c})`;
  return css({
    backgroundImage: isTo() ? [buildWebkitGradient(v as string, colorStops), o, g] : [o, g],
  });
}

export function buildLinearGradient(direction: CSSSideOrCorner | CSSAngle | string, ...colorStops: CSSLinearColorStopOrHint[]) {
  return buildGradientDirection(
    typeof direction === "string" && direction.startsWith("to ") ? direction.slice(3) : direction.toString(),
    colorStops
  )!;
}

export function buildGradientFrom(v: unknown) {
  return buildColor(prop`--w-gradient-from`, "--w-from-opacity", v, c => ({
    ...c,
    "--w-gradient-stops": "var(--w-gradient-from), var(--w-gradient-to, rgba(255, 255, 255, 0))",
  }));
}

export function buildGradientVia(v: unknown) {
  return buildColor(prop`--w-gradient-via`, "--w-via-opacity", v, c => ({
    "--w-gradient-stops": `var(--w-gradient-from), ${c["--w-gradient-via"]}, var(--w-gradient-to, rgba(255, 255, 255, 0))`,
  }));
}

export function buildGradientTo(v: unknown) {
  return buildColor(prop`--w-gradient-to`, "--w-to-opacity", v);
}

export function buildDivideColor(v: unknown) {
  return buildColor("borderColor", "--w-divide-opacity", v, prefixNotHidden);
}

export function buildDivideOpacity(v: unknown) {
  return buildNotHidden(prop`--w-divide-opacity`)(v);
}

export function buildDivideStyle(v: unknown) {
  return buildNotHidden("borderStyle")(v);
}

export function buildRingWidth(v: unknown) {
  return css({
    "--w-ring-offset-shadow": "var(--w-ring-inset) 0 0 0 var(--w-ring-offset-width) var(--w-ring-offset-color)",
    "--w-ring-shadow": `var(--w-ring-inset) 0 0 0 calc(${v} + var(--w-ring-offset-width)) var(--w-ring-color)`,
    "-webkit-box-shadow": "var(--w-ring-offset-shadow), var(--w-ring-shadow), var(--w-shadow, 0 0 #0000)",
    boxShadow: "var(--w-ring-offset-shadow), var(--w-ring-shadow), var(--w-shadow, 0 0 #0000)",
  });
}

export function buildBoxShadowSize(v: unknown) {
  return css({
    "--w-shadow": v as string,
    "--w-shadow-colored": (v as string).replace(/rgba?\([0-9.,/\s]*\)/g, "var(--w-shadow-color)"),
    "-webkit-box-shadow": "var(--w-ring-offset-shadow, 0 0 #0000), var(--w-ring-shadow, 0 0 #0000), var(--w-shadow)",
    boxShadow: "var(--w-ring-offset-shadow, 0 0 #0000), var(--w-ring-shadow, 0 0 #0000), var(--w-shadow)",
  });
}

export function buildBoxShadowColor(v: unknown) {
  return buildColor(prop`--w-shadow-color`, "--w-shadow-color-opacity", v, o => ({
    ...o,
    "--w-shadow": "var(--w-shadow-colored)",
  }));
}

export function buildImageRendering(v: unknown) {
  return css(prefixImageRendering(buildProperty("imageRendering", v as string)));
}

export function buildPlaceholder(v: unknown) {
  return buildColor("color", "--w-placeholder-opacity", v, prefixPlaceholder);
}

export function buildWritingMode(v: unknown) {
  return css(prefixWritingMode(buildProperty("writingMode", v as string)));
}
