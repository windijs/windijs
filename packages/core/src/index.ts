export * from "./base";
export * from "./api";
export * from "./handler";
export * from "./prefixer";
export * from "./builder";
