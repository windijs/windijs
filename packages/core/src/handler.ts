import { CSSMap, CSSObject, StyleObject, prop } from "@windi/helpers";
import { buildFontSize, buildKeyframes, buildProperty } from "./builder";
import { configHandler, cssHandler, genericHandler, guard } from "./api";
import { prefixAnimation, prefixNotHidden } from "./prefixer";

import { isNumber } from "@windi/shared";

export const animateHandler = <T extends string> (name: T, value: string | CSSObject, keyframes?: Record<string, CSSObject>) => {
  const map: CSSMap = new Map();
  const setValue = (o: CSSObject) => Object.entries(o).map(([k, v]) => map.set(k, v));
  if (keyframes) setValue(buildKeyframes(name, keyframes));
  setValue(typeof value === "string" ? buildProperty([prop`-webkit-animation`, "animation"], value) : prefixAnimation(value));

  return guard(name, cssHandler(map));
};

export function backgroundGenericHandler () {
  return genericHandler<{ [key: string]: StyleObject } & { $: {[key: string]: StyleObject} }>("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}

export function divideXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-x-reverse": "1",
  }));
}

export function divideYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-y-reverse": "1",
  }));
}

export function fontFamilyHandler<T extends object> (fonts: T) {
  const cssFonts = {} as {[key in keyof typeof fonts]: string};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key as keyof typeof fonts] = Array.isArray(value) ? value.join(",") : value as string;
  }
  return configHandler(cssFonts as T, "fontFamily");
}

export function fontSizeHandler<T extends object> (sizes: T) {
  return configHandler(sizes, value => {
    if (typeof value === "string") return buildFontSize(value);
    if (Array.isArray(value) && value[0]) {
      if (value[1] == null || typeof value[1] === "string") return buildFontSize(value[0], value[1]);
      if (typeof value[1] === "object" && value[1] != null) return buildFontSize(value[0], undefined, value[1]);
    }
  });
}

export function spaceBetweenXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-x-reverse": "1",
  }));
}

export function spaceBetweenYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-y-reverse": "1",
  }));
}
