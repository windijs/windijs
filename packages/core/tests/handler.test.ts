import { StyleObject, css, rgb } from "@windijs/helpers";
import { backgroundClipConfig, backgroundRepeatConfig, backgroundSizeConfig, borderStyleConfig, borderWidthConfig, fontStyleConfig, justifyItemsConfig, overflowConfig } from "@windijs/config";
import { backgroundGenericHandler, callHandler, colorHandler, configHandler, createUtility, genericHandler, guard, meld, setup, setupHandler, setupUtility, use } from "../src";
import { isNumber, parenWrap } from "@windijs/shared";

import { colors } from "@windijs/utilities";

test("useGeneric With Trigger", () => {
  const backgroundGeneric = genericHandler("backgroundColor", prop => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });

  const bg = createUtility("bg")
    .case("trigger", backgroundGeneric)
    .init();

  expect(bg.trigger["rgb(22, 22, 22)"].css).toMatchSnapshot();
});

test("Call Handler", () => {
  const rgbCSS: (red: number, green: number, blue: number) => StyleObject = (red: number, green: number, blue: number) => css({ backgroundColor: rgb(red, green, blue) });
  Object.defineProperty(rgbCSS, "abc", { value: rgbCSS });

  const bg = createUtility("bg")
    .use(colorHandler(colors, "backgroundColor"))
    .case("rgb", callHandler(
      rgbCSS as typeof rgbCSS & { abc: typeof rgbCSS },
      genericHandler("backgroundColor", v => parenWrap("rgb", v)),
    ),
    ).init();
  expect(bg.red[500].css).toMatchSnapshot();
  expect(bg.rgb(22, 22, 22).css).toMatchSnapshot();
  expect(bg.rgb(22, 22, 22).meta).toMatchSnapshot();
  expect(bg.rgb.abc(22, 22, 22).css).toMatchSnapshot();
  expect(bg.rgb["22, 22, 22"].css).toMatchSnapshot();
});

test("guard with meld", () => {
  const border = createUtility("border")
    .use(guard("colors", guard("dark", colorHandler(colors, "borderColor", "--w-border-opacity"))))
    .use(guard("width", meld(configHandler(borderWidthConfig, "borderWidth"), colorHandler(colors, "borderColor", "--w-border-opacity"))))
    .init();

  expect(border.colors.dark.amber[100].css).toMatchSnapshot();
  expect(border.width[2].css).toMatchSnapshot();
  expect(border.width.blue).toBeDefined();
  expect(border.width.blue[400].css).toMatchSnapshot();
});

test("two guard with same name should not work", () => {
  const border = createUtility("border")
    .use(guard("width", configHandler(borderWidthConfig, "borderWidth")))
    .use(guard("width", colorHandler(colors, "borderColor", "--w-border-opacity")))
    .init();

  expect(border.width[2]).toBeDefined();
  expect(border.width.blue).toBeUndefined();
});

test("use single plugin", () => {
  const bg = use("bg", backgroundGenericHandler());

  expect(bg["rgb(22, 22, 22)"].css).toMatchSnapshot();
  expect(bg.red.css).toMatchSnapshot();
  expect(bg.aliceBlue.css).toMatchSnapshot();
  expect(bg[0xffffff].css).toMatchSnapshot();
});

test("use multi plugin with meld and use", () => {
  const border = use("border", meld(configHandler(borderWidthConfig, "borderWidth"), colorHandler(colors, "borderColor", "--w-border-opacity")));
  const border2 = use("border", meld(configHandler(borderWidthConfig, "borderWidth"), colorHandler(colors, "borderColor", "--w-border-opacity"), configHandler(borderStyleConfig, "borderStyle")));

  expect(border[0].css).toMatchSnapshot();
  expect(border.emerald[500].css).toMatchSnapshot();
  expect(border2[4].css).toEqual(border[4].css);
  expect(border2.emerald[500].css).toEqual(border.emerald[500].css);
  expect(border2.dashed.css).toMatchSnapshot();
});

test("setup", () => {
  const bg = setup({
    red: css({ backgroundColor: "red" }),
    size: configHandler(backgroundSizeConfig, "backgroundSize"),
    repeat: configHandler(backgroundRepeatConfig.repeat, "backgroundRepeat"),
    multi: meld(
      configHandler(justifyItemsConfig, "justifyItems"),
      configHandler(fontStyleConfig, "fontStyle"),
    ),
    nested: {
      blue: css({ backgroundColor: "blue" }),
      opacity: {
        10: css({ backgroundColor: "aqua" }),
      },
      size: configHandler(backgroundSizeConfig, "backgroundSize"),
    },
  });
  expect(bg.red.css).toMatchSnapshot();
  expect(bg.size.auto.css).toMatchSnapshot();
  expect(bg.repeat.css).toMatchSnapshot();
  expect(bg.repeat.x.css).toMatchSnapshot();
  expect(bg.multi.end.css).toMatchSnapshot();
  expect(bg.multi.italic.css).toMatchSnapshot();
  expect(bg.nested.blue.css).toMatchSnapshot();
  expect(bg.nested.size.auto.css).toMatchSnapshot();
  expect(bg.nested.opacity[10].css).toMatchSnapshot();
});

test("setupHandler", () => {
  const bg = createUtility("bg")
    .use(setupHandler({
      red: css({ backgroundColor: "red" }),
      size: configHandler(backgroundSizeConfig, "backgroundSize"),
    }))
    .use(configHandler(backgroundRepeatConfig, "backgroundRepeat"))
    .init();

  expect(bg.red.css).toMatchSnapshot();
  expect(bg.size.auto.css).toMatchSnapshot();
  expect(bg.repeat.css).toMatchSnapshot();
  expect(bg.repeat.x.css).toMatchSnapshot();
  expect(bg.noRepeat.css).toMatchSnapshot();

  expect(bg.size.auto.meta).toMatchSnapshot();
  expect(bg.repeat.x.meta).toMatchSnapshot();
});

test("setupUtility with css", () => {
  const red = setupUtility("red", css({ backgroundColor: "red" }));
  expect(red.css).toMatchSnapshot();
  expect(red.meta).toMatchSnapshot();
});

test("setupUtility with handler", () => {
  const bg = setupUtility("bg", configHandler(backgroundRepeatConfig, "backgroundRepeat"));
  expect(bg.repeat.css).toMatchSnapshot();
  expect(bg.noRepeat.css).toMatchSnapshot();
  expect(bg.repeat.x.css).toMatchSnapshot();
  expect(bg.repeat.meta).toMatchSnapshot();
  expect(bg.repeat.x.meta).toMatchSnapshot();
});

test("setupUtility with config", () => {
  const overflow = setupUtility("overflow", {
    DEFAULT: meld(
      configHandler(overflowConfig, "overflow"),
      configHandler(backgroundClipConfig, "overflow"),
    ),
    ellpsis: css({ "-o-text-overflow": "ellipsis", textOverflow: "ellipsis" }),
    x: configHandler(overflowConfig, "overflowX"),
    multi: meld(
      configHandler(justifyItemsConfig, "justifyItems"),
      configHandler(fontStyleConfig, "fontStyle"),
    ),
    nested: {
      blue: css({ backgroundColor: "blue" }),
      opacity: {
        10: css({ backgroundColor: "aqua" }),
      },
      size: configHandler(backgroundSizeConfig, "backgroundSize"),
    },
  });

  expect(overflow.border.css).toMatchSnapshot();
  expect(overflow.clip.css).toMatchSnapshot();
  expect(overflow.ellpsis.css).toMatchSnapshot();
  expect(overflow.x.clip.css).toMatchSnapshot();
  expect(overflow.multi.center.css).toMatchSnapshot();
  expect(overflow.multi.italic.css).toMatchSnapshot();
  expect(overflow.nested.blue.css).toMatchSnapshot();
  expect(overflow.nested.opacity[10].css).toMatchSnapshot();
  expect(overflow.nested.size.contain.css).toMatchSnapshot();

  expect(overflow.border.meta).toMatchSnapshot();
  expect(overflow.clip.meta).toMatchSnapshot();
  expect(overflow.ellpsis.meta).toMatchSnapshot();
  expect(overflow.x.clip.meta).toMatchSnapshot();
  expect(overflow.multi.center.meta).toMatchSnapshot();
  expect(overflow.multi.italic.meta).toMatchSnapshot();
  expect(overflow.nested.blue.meta).toMatchSnapshot();
  expect(overflow.nested.opacity[10].meta).toMatchSnapshot();
  expect(overflow.nested.size.contain.meta).toMatchSnapshot();
});
