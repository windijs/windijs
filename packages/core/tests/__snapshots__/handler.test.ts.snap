// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Call Handler 1`] = `
Object {
  "backgroundColor": "#ef4444",
}
`;

exports[`Call Handler 2`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;

exports[`Call Handler 3`] = `
Object {
  "props": Array [
    "rgb(22,22,22)",
  ],
  "type": "color",
  "uid": "bg",
  "variants": Array [],
}
`;

exports[`Call Handler 4`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;

exports[`Call Handler 5`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;

exports[`guard with meld 1`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(254, 243, 199, var(--w-border-opacity))",
}
`;

exports[`guard with meld 2`] = `
Object {
  "borderWidth": "2px",
}
`;

exports[`guard with meld 3`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(96, 165, 250, var(--w-border-opacity))",
}
`;

exports[`setup 1`] = `
Object {
  "backgroundColor": "red",
}
`;

exports[`setup 2`] = `
Object {
  "backgroundSize": "auto",
}
`;

exports[`setup 3`] = `
Object {
  "backgroundRepeat": "repeat",
}
`;

exports[`setup 4`] = `
Object {
  "backgroundRepeat": "repeat-x",
}
`;

exports[`setup 5`] = `
Object {
  "justifyItems": "end",
}
`;

exports[`setup 6`] = `
Object {
  "fontStyle": "italic",
}
`;

exports[`setup 7`] = `
Object {
  "backgroundColor": "blue",
}
`;

exports[`setup 8`] = `
Object {
  "backgroundSize": "auto",
}
`;

exports[`setup 9`] = `
Object {
  "backgroundColor": "aqua",
}
`;

exports[`setupHandler 1`] = `
Object {
  "backgroundColor": "red",
}
`;

exports[`setupHandler 2`] = `
Object {
  "backgroundSize": "auto",
}
`;

exports[`setupHandler 3`] = `
Object {
  "backgroundRepeat": "repeat",
}
`;

exports[`setupHandler 4`] = `
Object {
  "backgroundRepeat": "repeat-x",
}
`;

exports[`setupHandler 5`] = `
Object {
  "backgroundRepeat": "no-repeat",
}
`;

exports[`setupHandler 6`] = `
Object {
  "props": Array [
    "size",
    "auto",
  ],
  "type": "static",
  "uid": "bg",
  "variants": Array [],
}
`;

exports[`setupHandler 7`] = `
Object {
  "props": Array [
    "repeat",
    "x",
  ],
  "type": "static",
  "uid": "bg",
  "variants": Array [],
}
`;

exports[`setupUtility with config 1`] = `
Object {
  "overflow": "border-box",
}
`;

exports[`setupUtility with config 2`] = `
Object {
  "overflow": "clip",
}
`;

exports[`setupUtility with config 3`] = `
Object {
  "-o-text-overflow": "ellipsis",
  "textOverflow": "ellipsis",
}
`;

exports[`setupUtility with config 4`] = `
Object {
  "overflowX": "clip",
}
`;

exports[`setupUtility with config 5`] = `
Object {
  "justifyItems": "center",
}
`;

exports[`setupUtility with config 6`] = `
Object {
  "fontStyle": "italic",
}
`;

exports[`setupUtility with config 7`] = `
Object {
  "backgroundColor": "blue",
}
`;

exports[`setupUtility with config 8`] = `
Object {
  "backgroundColor": "aqua",
}
`;

exports[`setupUtility with config 9`] = `
Object {
  "backgroundSize": "contain",
}
`;

exports[`setupUtility with config 10`] = `
Object {
  "props": Array [
    "border",
  ],
  "type": "static",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 11`] = `
Object {
  "props": Array [
    "clip",
  ],
  "type": "static",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 12`] = `
Object {
  "props": Array [
    "ellpsis",
  ],
  "type": "css",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 13`] = `
Object {
  "props": Array [
    "x",
    "clip",
  ],
  "type": "static",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 14`] = `
Object {
  "props": Array [
    "multi",
    "center",
  ],
  "type": "static",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 15`] = `
Object {
  "props": Array [
    "multi",
    "italic",
  ],
  "type": "static",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 16`] = `
Object {
  "props": Array [
    "nested",
    "blue",
  ],
  "type": "css",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 17`] = `
Object {
  "props": Array [
    "nested",
    "opacity",
    "10",
  ],
  "type": "css",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with config 18`] = `
Object {
  "props": Array [
    "nested",
    "size",
    "contain",
  ],
  "type": "static",
  "uid": "overflow",
  "variants": Array [],
}
`;

exports[`setupUtility with css 1`] = `
Object {
  "backgroundColor": "red",
}
`;

exports[`setupUtility with css 2`] = `
Object {
  "props": Array [],
  "type": "css",
  "uid": "red",
  "variants": Array [],
}
`;

exports[`setupUtility with handler 1`] = `
Object {
  "backgroundRepeat": "repeat",
}
`;

exports[`setupUtility with handler 2`] = `
Object {
  "backgroundRepeat": "no-repeat",
}
`;

exports[`setupUtility with handler 3`] = `
Object {
  "backgroundRepeat": "repeat-x",
}
`;

exports[`setupUtility with handler 4`] = `
Object {
  "props": Array [
    "repeat",
  ],
  "type": "static",
  "uid": "bg",
  "variants": Array [],
}
`;

exports[`setupUtility with handler 5`] = `
Object {
  "props": Array [
    "repeat",
    "x",
  ],
  "type": "static",
  "uid": "bg",
  "variants": Array [],
}
`;

exports[`use multi plugin with meld and use 1`] = `
Object {
  "borderWidth": "0px",
}
`;

exports[`use multi plugin with meld and use 2`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(16, 185, 129, var(--w-border-opacity))",
}
`;

exports[`use multi plugin with meld and use 3`] = `
Object {
  "borderStyle": "dashed",
}
`;

exports[`use single plugin 1`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;

exports[`use single plugin 2`] = `
Object {
  "backgroundColor": "red",
}
`;

exports[`use single plugin 3`] = `
Object {
  "backgroundColor": "aliceBlue",
}
`;

exports[`use single plugin 4`] = `
Object {
  "backgroundColor": "#ffffff",
}
`;

exports[`useGeneric With Trigger 1`] = `
Object {
  "backgroundColor": "rgb(22, 22, 22)",
}
`;
